src/emitter/batch_emitter.rs

```diff
@@ -163,43 +235,103 @@ impl BatchEmitter {
-                let message = match rx.recv().await {
-                    Some(msg) => {
-                        log::debug!("Received message: {:?}", msg);
-                        msg
-                    }
+
+                // select! is used to check both the `retry_rx` channel and the `rx` channel for new messages
+                let message = match tokio::select! {
+                    // `biased;` is used to ensure that the `retry_rx` channel is checked first, so retries get priority
+                    biased;
+
+                    retry = retry_rx.recv() => retry,
+                    event = rx.recv() => event,
+                } {
+                    Some(message) => message,
                     None => break,
                 };
 
                 match message {
-                    EmitterMessage::Send(batch) => {
-                        // Clone HttpClient to be moved into the task
+                    EmitterMessage::Send(mut batch) => {
+                        // Clone to move into the task
                         let client = http_client.clone();
+                        let retry_transmitter = retry_tx.clone();
+                        let store = event_store.clone();
 
                         // Spawn a new task to send the batch
                         tokio_tasks.push(tokio::spawn(async move {
+                            if let Some(delay) = batch.delay {
+                                log::debug!("Delaying batch {} for {:?}", batch.id, delay);
+                                tokio::time::sleep(delay).await;
+
+                                if let Err(e) = batch.update_event_stm() {
+                                    // If the update fails, we just re-send the batch as-is
+                                    // Not ideal, but it's better than losing events
+                                    log::warn!(
+                                        "Failed to update stm of events in batch {} for retry: {e}",
+                                        batch.id
+                                    )
+                                };
+                            };
+
                             let batch_length = batch.events.len();
                             match Self::send_batch(batch, client).await {
-                                Ok(_) => {
-                                    log::info!("Sent batch of {batch_length} events")
+                                Ok(resp) => {
+                                    // We got a response from the collector, but need to check if
+                                    // it was successful
+                                    if Self::should_retry(resp.code) && resp.batch.has_retry() {
+                                        Self::retry_batch(resp.batch, retry_transmitter)
+                                    } else {
+                                        log::info!(
+                                            "Sent batch {} of {batch_length} events",
```

Comment:
Nit: this message might be misleading since we can get here also in case request failed but we are not allowed to retry, right?

Reply:
I've re-worked the logic a bit on the `Ok` of `send_batch` to log this when appropriate.